go/teams/transactions.go

```diff
@@ -222,6 +222,64 @@ func (tx *AddMemberTx) AddMemberByUsername(ctx context.Context, username string,
 	return tx.addMember(uv, role)
 }
 
+// AddMemberByUsername will add member by UV and role. It checks if
+// given UV is valid (that we don't have outdated EldestSeqno), and if
+// user has PUK, and if not, it properly handles that by adding
+// Keybase-type invite. It also cleans up old invites and memberships.
+func (tx *AddMemberTx) AddMemberByUV(ctx context.Context, uv keybase1.UserVersion, role keybase1.TeamRole) (err error) {
+	team := tx.team
+	g := team.G()
+
+	defer g.CTrace(ctx, fmt.Sprintf("AddMemberTx.AddMemberByUV(%v,%v) to team %q", uv, role, team.Name()), func() error { return err })()
+	upak, err := loadUPAK2(ctx, g, uv.Uid, true /*forcePoll */)
+	if err != nil {
+		return err
+	}
+
+	current := upak.Current
+	if uv.EldestSeqno != current.EldestSeqno {
+		return fmt.Errorf("Bad eldestseqno for %s: expected %d, got %d", uv.Uid, current.EldestSeqno, uv.EldestSeqno)
+	}
+
+	return tx.addMemberByUPKV2(ctx, current, role)
+}
+
+// AddMemberByUsername will add member by username and role. It
+// checks if given username can become crypto member or a PUKless
+// member. It will also clean up old invites and memberships if
+// necessary.
+func (tx *AddMemberTx) AddMemberByUsername(ctx context.Context, username string, role keybase1.TeamRole) (err error) {
+	team := tx.team
+	g := team.G()
+
+	defer g.CTrace(ctx, fmt.Sprintf("AddMemberTx.AddMemberByUsername(%s,%v) to team %q", username, role, team.Name()), func() error { return err })()
+
+	res := g.Resolver.ResolveWithBody(username)
+	if err := res.GetError(); err != nil {
+		return err
+	}
+
+	upak, err := loadUPAK2(ctx, g, res.GetUID(), true /* forcePoll */)
+	if err != nil {
+		return err
+	}
+
+	return tx.addMemberByUPKV2(ctx, upak.Current, role)
+}
+
+func (tx *AddMemberTx) CompleteInviteByID(ctx context.Context, inviteID keybase1.TeamInviteID, uv keybase1.UserVersion) error {
```

Comment:
Should this be uncapitalized? I'd guess you only complete an invite in lockstep with having them in the members section of a link.

Reply:
I'm going to merge this as is, I have a couple more PRs for integrating team transactions to various places and the API is a bit of a moving target.